src/odemis/driver/technolution.py

```diff
@@ -0,0 +1,965 @@
+        if delay >= self.parent._ebeam_scanner.scanDelay.value[0]:
+            return delay
+        else:
+            # Change values so that 'self.acqDelay.value - self.parent._ebeam_scanner.scanDelay.value[0]' has a positive result
+            logging.warning("Detector cannot record images before ebeam scanner has started to scan.\n"
+                            "Detector needs to start after scanner.")
+            logging.info("The entered acquisition delay is %s in the eBeamScanner and the scan delay in the MPPC is "
+                         "%s" % (delay, self.parent._ebeam_scanner.scanDelay.value[0]))
+            return self.acqDelay.value
+
+    def _setFilename(self, file_name):
+        """
+        Check if filename complies with set allowed characters
+        :param file_name:
+        :return:
+        """
+        ASM_FILE_ALLOWED_CHARS = r'[^a-z0-9_()-]'
+        search = re.compile(ASM_FILE_ALLOWED_CHARS).search
+        if not bool(search(file_name)):
+            return file_name
+        else:
+            logging.warning("File_name contains invalid characters, file_name remains unchanged (only the characters "
+                            "'%s' are allowed)" % ASM_FILE_ALLOWED_CHARS[2:-1])
+            return self.filename.value
+
+    def _setURL(self, url_parser):
+        """
+        Setter which checks for correctness of FTP url_parser and otherwise returns old value.
+
+        :param url_parser: e.g. ftp://username:password@example.com
+        :return: correct ftp url_parser
+        """
+        ASM_GENERAL_ALLOWED_CHARS = r'[^A-Za-z0-9/_()-:@]'
+        ASM_USER_ALLOWED_CHARS = r'[^A-Za-z0-9]'
+        ASM_PASSWORD_ALLOWED_CHARS = r'[^A-Za-z0-9]'
+        ASM_HOST_ALLOWED_CHARS = r'[^A-Za-z0-9.]'
+        ASM_PATH_ALLOWED_CHARS = r'[^A-Za-z0-9/_()-]'
+
+        def checkCharacters(input, allowed_characters):
+            """
+            Check if input complies with allowed characters
+            :param input (sting): input string
+            :param allowed_characters: allowed_characters for different parts of input string
+            :return (boolean) True if passes test on allowed_characters
+            """
+            search = re.compile(allowed_characters).search
+            if not bool(search(input)):
+                return True
+            else:
+                return False
+
+        # Perform general check on valid characters (parses works incorrectly for some invalid characters
+        if not checkCharacters(urlunparse(url_parser), ASM_GENERAL_ALLOWED_CHARS):
+            logging.warning("Invalid character in ftp url is provided, allowed characters are %s in the form:: "
+                            "'ftp://username:password@host_example.com/path/to/Pictures'\n"
+                            "(Only use the @ to separate the password and the host." % ASM_GENERAL_ALLOWED_CHARS[2:-1])
+            return self.externalStorageURL.value
```

Comment:
I think checking for this is clearer for the user because if any weird characters are present in the input string it actually won't be possible to break down the string. For example # can mess the entire parsing of the url which result in for example to having a username in the output.
Therefore I have this general check also implemented because otherwise the other checks will fail and give an unclear error.

Reply:
If there are odd characters in the URL the URL cannot be parsed correctly. For example a '#' value can result in not having a username path etc in the parsed object. This would otherwise result in an unclear error message. I think that this implementation could at least catch this specific case and present a clearer message.
Or do you think that is not needed and such a mistake is rare?

Reply:
@pieleric this method moved to a different class (Acquisition server)

Reply:
Alright, indeed, in case of error, that can make it harder to decode. Just leave it as-is (but be sure to not be too restrictive in the forbidden characters).